### PR TITLE
aarch64: Fix compilation for windows-on-arm (msys2)

### PR DESCRIPTION
**Pull request by kd-11**
Untested as the build doesn't run, but an important step nonetheless.

### CODE REVIEW

File: rpcs3/Emu/CPU/sse2neon.h [hunk 1]
```diff
@@ -132,6 +132,25 @@
 #include <math.h>
 #endif
 
+#if defined(_WIN32)
+static inline int posix_memalign(void** memptr, size_t alignment, size_t size) {
+    // Check for valid alignment (must be a power of two and multiple of sizeof(void*))
+    if (alignment == 0 || (alignment & (alignment - 1)) != 0 || alignment % sizeof(void*) != 0) {
```
Comment (Megamouse) [inline rpcs3/Emu/CPU/sse2neon.h:138]: code style (braces)

File: rpcs3/Emu/CPU/sse2neon.h [hunk 2]
```diff
@@ -132,6 +132,25 @@
...
+        return EINVAL;
+    }
+
+    // Allocate memory using _aligned_malloc
+    void* ptr = _aligned_malloc(size, alignment);
+    if (ptr == NULL) {
```
Comment (Megamouse) [inline rpcs3/Emu/CPU/sse2neon.h:144]: code style (braces)
nullptr ?

File: rpcs3/Emu/CPU/sse2neon.h [hunk 3]
```diff
@@ -132,6 +132,25 @@
 #include <math.h>
 #endif
 
+#if defined(_WIN32)
+static inline int posix_memalign(void** memptr, size_t alignment, size_t size) {
```
Comment (Megamouse) [inline rpcs3/Emu/CPU/sse2neon.h:136]: code style (braces)

File: Utilities/Thread.h [hunk 4]
```diff
@@ -510,15 +510,19 @@ class named_thread final : public Context, result_storage<Context>, thread_base
...
 	static void* trampoline(void* arg)
+#endif
 	{
 		if (const auto next = thread_base::finalize(entry_point(static_cast<thread_base*>(arg))))
 		{
 			return next(thread_ctrl::get_current());
 		}
 
-		return nullptr;
+		return 0;
```
Comment (Megamouse) [inline Utilities/Thread.h:525]:
```suggestion
		return {};
```

**Comment by oltolm** [comment]
I get this warning when building

```
[4/9] Building CXX object rpcs3/Emu/CMakeFiles/rpcs3_emu.dir/__/__/Utilities/Thread.cpp.obj
C:/src/rpcs3/Utilities/Thread.cpp:1844:9: warning: 'RIP' macro redefined [-Wmacro-redefined]
 1844 | #define RIP(ctx) ctx->Rip
      |         ^
C:/src/rpcs3/Utilities/Thread.cpp:1026:9: note: previous definition is here
 1026 | #define RIP(c) (*X64REG((c), 16))
      |         ^
1 warning generated.
```

**Comment by Darkhost1999** [comment]
Looks like it's being worked on and merged into a larger PR
https://github.com/RPCS3/rpcs3/pull/16046

**Comment by oltolm** [comment]
It also broke building with GCC on Windows.
```
[601/604] Linking CXX executable bin\rpcs3.exe
FAILED: bin/rpcs3.exe
C:\WINDOWS\system32\cmd.exe /C "cd . && C:\msys64\mingw64\bin\c++.exe -g -mwindows  -Wl,--allow-multiple-definition -Wl,--stack -Wl,8388608 -Wl,--image-base,0x10000 @CMakeFiles\rpcs3.rsp -o bin\rpcs3.exe -Wl,--out-implib,rpcs3\librpcs3.dll.a -Wl,--major-image-version,0,--minor-image-version,0 && C:\WINDOWS\system32\cmd.exe /C "cd /D E:\build-rpcs3-gcc\rpcs3 && C:\msys64\mingw64\bin\cmake.exe -E copy_directory C:/src/rpcs3/bin E:/build-rpcs3-gcc/bin && C:\msys64\mingw64\bin\windeployqt-qt6.exe --no-compiler-runtime --no-opengl-sw --no-patchqt --no-translations --no-quick --no-system-d3d-compiler --no-quick-import --plugindir E:/build-rpcs3-gcc/bin/share/qt6/plugins --verbose 0 E:/build-rpcs3-gcc/bin/rpcs3.exe""
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lc++: No such file or directory
collect2.exe: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

It's line 117 in `rpcs3/CMakeLists.txt`.

**Comment by kd-11** [comment]
Are gcc builds normally functional? Clang needs c++ lib on macos as well, so this is simply a clang vs gcc issue.

**Comment by oltolm** [comment]
Yes gcc builds are functional. I just tested it with Demon's Souls. The standard C++ library is normally specified outside of `CMakeLists.txt` in `CXX_FLAGS` or `CMAKE_CXX_FLAGS` like this `-stdlib=libc++`. See https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-stdlib.

**Comment by kd-11** [comment]
We'll not be relying on command-line switches for cmake, that's a no-go. We require the standard c++ library so it needs to be defined explicitly for clang.
I am working on a much bigger changeset for MSVC support, I'll fit in these minor thing with that one.

**Comment by oltolm** [comment]
BTW `libc++` is the default for the clang64 subsystem of msys2 on Windows. If you use clang64 then you do not need to specify `-stdlib`. [clangarm64](https://www.msys2.org/docs/environments/) also uses `libc++` per default.

**Pull request merged**